src/ray/core_worker/reference_count.h

```diff
@@ -126,22 +224,179 @@ class ReferenceCounter {
+    ///  2. A task that we submitted returned an ID(s).
+    /// ObjectIDs are erased from this field when their Reference is deleted.
+    absl::flat_hash_set<ObjectID> contained_in_owned;
+    /// An Object ID that we (or one of our children) borrowed that contains
+    /// this object ID, which is also borrowed. This is used in cases where an
+    /// ObjectID is nested. We need to notify the owner of the outer ID of any
+    /// borrowers of this object, so we keep this field around until
+    /// GetAndClearLocalBorrowersInternal is called on the outer ID. This field
+    /// is updated in 2 cases:
+    ///  1. We deserialize an ID that we do not own and that was stored in
+    ///     another object that we do not own.
+    ///  2. Case (1) occurred for a task that we submitted and we also do not
+    ///     own the inner or outer object. Then, we need to notify our caller
+    ///     that the task we submitted is a borrower for the inner ID.
+    /// This field is reset to null once GetAndClearLocalBorrowersInternal is
+    /// called on contained_in_borrowed_id. For each borrower, this field is
+    /// set at most once during the reference's lifetime. If the object ID is
+    /// later found to be nested in a second object, we do not need to remember
+    /// the second ID because we will already have notified the owner of the
+    /// first outer object about our reference.
+    absl::optional<ObjectID> contained_in_borrowed_id;
+    /// The object IDs contained in this object. These could be objects that we
+    /// own or are borrowing. This field is updated in 2 cases:
+    ///  1. We call ray.put() on this ID and store the contained IDs.
+    ///  2. We call ray.get() on an ID whose contents we do not know and we
+    ///     discover that it contains these IDs.
+    absl::flat_hash_set<ObjectID> contains;
+    /// A list of processes that are we gave a reference to that are still
+    /// borrowing the ID. This field is updated in 2 cases:
+    ///  1. If we are a borrower of the ID, then we add a process to this list
+    ///     if we passed that process a copy of the ID via task submission and
+    ///     the process is still using the ID by the time it finishes its task.
+    ///     Borrowers are removed from the list when we recursively merge our
+    ///     list into the owner.
+    ///  2. If we are the owner of the ID, then either the above case, or when
+    ///     we hear from a borrower that it has passed the ID to other
+    ///     borrowers. A borrower is removed from the list when it responds
+    ///     that it is no longer using the reference.
+    absl::flat_hash_set<rpc::WorkerAddress> borrowers;
+
+    /// Callback that will be called when this ObjectID no longer has
+    /// references.
     std::function<void(const ObjectID &)> on_delete;
+    /// Callback that is called when this process is no longer a borrower
+    /// (RefCount() == 0).
+    std::function<void()> on_local_ref_deleted;
```

Comment:
The name of this is confusing because `local_ref` makes me think it will be called when local Python references go out of scope.